databases/nmdc/database.go

```diff
@@ -0,0 +1,992 @@
+		if err != nil {
+			return nil, err
+		}
+		var dataObject DataObject
+		err = json.Unmarshal(body, &dataObject)
+		if err != nil {
+			return nil, err
+		}
+		resources[i], err = db.dataResourceFromDataObject(dataObject)
+		if err != nil {
+			return nil, err
+		}
+
+		// add credit metadata
+		studyId := studyIdForDataObjectId[resources[i].Id]
+		resources[i].Credit = creditForStudyId[studyId]
+	}
+	return resources, nil
+}
+
+func (db Database) StageFiles(fileIds []string) (uuid.UUID, error) {
+	// NMDC keeps all of its NERSC data on disk, so all files are already staged.
+	// We simply generate a new UUID that can be handed to db.StagingStatus,
+	// which returns databases.StagingStatusSucceeded.
+	//
+	// "We may eventually use tape but don't need to yet." -Shreyas Cholia, 2024-09-04
+	return uuid.New(), nil
+}
+
+func (db Database) StagingStatus(id uuid.UUID) (databases.StagingStatus, error) {
+	// all files are hot!
+	return databases.StagingStatusSucceeded, nil
+}
+
+func (db Database) LocalUser(orcid string) (string, error) {
+	// no current mechanism for this
+	return "localuser", nil
+}
+
+//--------------------
+// Internal machinery
+//--------------------
+
+const (
+	// NOTE: for now, we use the dev environment (-dev), not prod (which has bugs!)
+	// NOTE: note also that NMDC is backed by two databases: one MongoDB and one PostGres,
+	// NOTE: which are synced daily-esque. They will sort this out in the coming year,
+	// NOTE: and it looks like PostGres is probably going to prevail.
+	// NOTE: (See https://github.com/microbiomedata/NMDC_documentation/blob/main/docs/howto_guides/portal_guide.md)
+	baseApiURL  = "https://api-dev.microbiomedata.org/"       // mongoDB
+	baseDataURL = "https://data-dev.microbiomedata.org/data/" // postgres (use in future)
+)
+
+// Authorization / authentication
+
+type authorization struct {
+	// API user credential
+	Credential credential
+	// client token and type (indicating how it's used in an auth header)
+	Token, Type string
+	// indicates whether the token expires
+	Expires bool
+	// time at which the token expires, if any
+	ExpirationTime time.Time
+}
+
+type credential struct {
+	User, Password string
+}
+
+// fetches an access token / type from NMDC using a credential
+func getAccessToken(credential credential) (authorization, error) {
+	var auth authorization
+	resource := baseApiURL + "token/"
+
+	// the token request must be URL-encoded
+	data := url.Values{}
+	data.Set("grant_type", "password")
+	data.Set("username", credential.User)
+	data.Set("password", credential.Password)
+	request, err := http.NewRequest(http.MethodPost, resource, strings.NewReader(data.Encode()))
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	request.Header.Set("Accept", "application/json")
+
+	var client http.Client
+	response, err := client.Do(request)
+	if err != nil {
+		return auth, err
+	}
+
+	switch response.StatusCode {
+	case 200, 201, 204:
```

Comment:
I'm assuming these success codes came from the NMDC devs

Reply:
Well, that and my experimenting. :-)